lib/shift_generator.rb

```diff
@@ -0,0 +1,25 @@
+require 'date'
+require './lib/key_generator'
+require './lib/offset_generator'
+
+module ShiftGenerator
+  def self.generate(key = nil, date = nil)
```

Comment:
set key and date equal to nil in the instance these are left empty. Memoization done below will generate a key or a date